File: sorting_test.rb
class User
  attr_reader :id, :name

  def initialize(id, name)
    @id = id
    @name = name
  end

  def inspect
    "User(#{id}, \"#{name}\")"
  end
end

if __FILE__ == $0
  users = [
    User.new(3, "A"),
    User.new(1, "B"),
    User.new(2, "C")
  ]
  puts "users: #{users}"

  sorted_users1 = users.sort { |x, y| x.id <=> y.id }
  puts "sorted_users1: #{sorted_users1}"

  sorted_users2 = users.sort_by { |x| x.id }
  puts "sorted_users2: #{sorted_users2}"

  sorted_users3 = users.sort_by(&:id)
  puts "sorted_users3: #{sorted_users3}"
end

File: SortingTest.hs
module SortingTest where

import Data.Function (on)
import Data.List (sortBy)

data User = User {
  id' :: Int,
  name :: String
} deriving (Eq, Show)

main :: IO ()
main = do
  putStrLn ("users: " ++ show users)
  putStrLn ("sortedUsers1: " ++ show sortedUsers1)
  putStrLn ("sortedUsers2: " ++ show sortedUsers2)
  where
    users = [User 3 "A", User 1 "B", User 2 "C"]
    sortedUsers1 = sortBy (\x y -> compare (id' x) (id' y)) users
    sortedUsers2 = sortBy (compare `on` id') users
